### PR TITLE
Parse environment markers from setup.py install_requires

### PR DESCRIPTION
**Pull request by iamhatesz**
**Changelog-friendly one-liner**: Added scanning of `extras_require` in `setup.py` for additional dependencies which might come from using environment markers in `install_requires`.

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by iamhatesz** [comment]
Hi @atugushev 

I am not sure if I understand #908 correctly. You propose the support of `extras_require` by passing `--extras` flag to `pip-compile` explicitly. This is fine when we have manually defined extras like `dev`, but I am afraid this could be a problem, when these extras are dynamically generated e.g. by specifying environment markers. In this case, I don't know what value should I use when calling `pip-compile` - maybe `:"linux" in sys_platform`, maybe `:platform_system == "Linux"`, too many cases.

**Comment by atugushev** [comment]
> I am not sure if I understand #908 correctly. You propose the support of `extras_require` by passing `--extras` flag to `pip-compile` explicitly. This is fine when we have manually defined extras like `dev`, but I am afraid this could be a problem, when these extras are dynamically generated e.g. by specifying environment markers. In this case, I don't know what value should I use when calling `pip-compile` - maybe `:"linux" in sys_platform`, maybe `:platform_system == "Linux"`, too many cases.

The original propose is to use `pip. install -e .` to parse `setup.py`.  To parse deps via `pip install -e .[my-extra]` (e.g., extra markers) proposed  `--extra=my-extra` option. You are right, it won't work for environment markers, but it makes no sense since, for example, it's difficult to build Windows dependencies on a Linux platform.

**Comment by iamhatesz** [comment]
Hm, I don't consider parsing Windows-specific dependencies when being on Linux. I am fine with compiling dependencies twice: once on a Windows machine, and the second time on a Linux machine. 

In my project, I have a few dependencies that can be satisfied only on Linux. Thus in my `setup.py` I use environment markers, so installing my package from `pip` works well. However, with current `pip-tools` I can't successfully compile my `setup.py` to `requirements.txt`, because it doesn't parse `extra_requires`. The solution with `--extras` flag is a workaround, but in this case I have to know what value I have to provide there. And for some cases, I can't be sure that, because some other packages could use environment markers as well, and I don't know to what exact extras key they will be transformed.

**Comment by atugushev** [comment]
@iamhatesz 

> If `pip install -e .` parses extras which come from environment markers (satisfying only these appropriate for current platform), then this is exactly what I'd like to achieve.

Yes, it works for the environment markers. You can try it how it works by:

```
$ echo "-e ." | pip-compile - -qo- | sed '/^-e / d'
```

**Comment by iamhatesz** [comment]
@atugushev cool, just tested, seems like a solution for me. What's the status of #908 then? Can I help you with the implementation?

**Comment by atugushev** [comment]
@iamhatesz 

> cool, just tested, seems like a solution for me. What's the status of #908 then? 

It's still under discussion, you might join there and express your thoughts.

> Can I help you with the implementation?

Sure, any contributions are highly welcomed and encouraged 😉

File: tests/test_cli_compile.py [hunk 1]
```diff
@@ -58,6 +58,36 @@ def test_command_line_setuptools_read(pip_conf, runner):
...
 
+def test_command_line_setuptools_with_env_markers_read(pip_conf, runner):
+    package = open("setup.py", "w")
+    package.write(
+        dedent(
+            """\
+            from setuptools import setup
+            setup(
+                install_requires=[
+                    'small-fake-a==0.1 ; "linux" in sys_platform',
```
Comment (webknjaz) [inline tests/test_cli_compile.py:69]: I'd add some corner cases. Like empty or whitespace-only string after `;`, also multiple conditions

**Comment by anthrotype** [comment]
Hi. Any updates on this? I'd love to see support for environment markers integrated in `pip-compile setup.py`. Thank you

**Comment by atugushev** [comment]
I'm going to close this in favor of #1311. Thanks for the PR anyway! 🙏🏼
